### PR TITLE
Support correlated column mapping type coercion

### LINKED ISSUE #15574: Support to handle type coercion for correlated subquery
When correlated symbols are in non-direct column reference type, query is failed..

presto:test> select (select count(*) from (values 1) t(a) where t.a=t2.b group by a limit 1) from (values 1.0) t2(b);

`Query 20201230_194256_00000_ydmqh failed: line 1:9: Given correlated subquery is not supported`

### PR DESCRIPTION
**Pull request by fgwang7w**
This PR adds type coercion support when correlated symbols are in non-direct column reference type.

Test plan -
Test in local, add UT coverage

resolve #15574 

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java [hunk 1]
```diff
@@ -89,13 +89,13 @@ public void testCorrelatedExistsSubqueriesWithOrPredicateAndNull()
     public void testUnsupportedSubqueriesWithCoercions()
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:89]: rename this test `testSubqueriesWithCoercions` since this is supported now

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java [hunk 2]
```diff
@@ -111,7 +112,7 @@ public Result apply(LateralJoinNode lateralJoinNode, Captures captures, Context
     {
         return searchFrom(rootNode, lookup)
                 .where(AggregationNode.class::isInstance)
-                .recurseOnlyWhen(isInstanceOfAny(ProjectNode.class, EnforceSingleRowNode.class))
+                .recurseOnlyWhen(isInstanceOfAny(ProjectNode.class, EnforceSingleRowNode.class, LimitNode.class))
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java:115]: This looks unrelated to the coercion issue.  Can you put it in a separate commit (can be in the same pr) and also explain the change.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 3]
```diff
@@ -277,29 +278,52 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
...
                 }
 
                 ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                if (!(comparison.getLeft() instanceof SymbolReference
-                        && comparison.getRight() instanceof SymbolReference
-                        && comparison.getOperator().equals(EQUAL))) {
+                // handle coercions and non-direct column references
+                Expression leftExpression;
+                Expression rightExpression;
+                if (comparison.getLeft() instanceof Cast) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:291]: won't it be a problem that we're getting rid of the casts if they are needed?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 4]
```diff
@@ -277,29 +278,52 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
 
         private Multimap<VariableReferenceExpression, VariableReferenceExpression> extractCorrelatedSymbolsMapping(List<Expression> correlatedConjuncts)
         {
-            // TODO: handle coercions and non-direct column references
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:280]: I'm not sure what non-direct column references is specifically, but I'm suspecting it's if you had some kind of operation on the column, e.g t1.a + 1 = t2.a instead of just t1.a = t2.a. I guess that is still unsupported.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java (hunk 2 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedScalarAggregationToJoin.java:115]: when the subquery ends with a LIMIT clause, the transformation needs to go one further step to look for aggNode because limit could be part of the correlated subquery above the agg. sure will place in a separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 3 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:291]: not clear what you meant here. The logic is to extract both left and right sides of the comparsion expression if it is formed of CAST expressions in order to find the symbol mapping. If there's no CAST expression, then the logic goes BAU to check for symbolReference Type, or to skip to the next iteration.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 4 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:280]: yes I was planning to do it separatelly. Non-direct column references like you point out may have different expression type like ArithmeticBinaryExpression, or `t1.a = -t2.a` for ArithmeticUnaryExpression, or `t1.a != t2.a` for NotExpression. For now I think euqality expression is the most common use case so I am adding this feature within the commit. I will add this comment as a TODO later

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 1 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 3 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:292]: What I mean is that this result is then used as the filter for a join https://github.com/prestodb/presto/blob/0df58706580bc9803ab0a33354d47e25ae63b372/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedLateralJoinToJoin.java#L66.  Couldn't the missing cast be an issue?

Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:292]: actually there is no changes to the correlated predicate in `DecorrelatedResult`. The logic is to extract the symbol of column reference from each side of predicate and create a direct variable mapping. As a decorrelated result, the original predicate still contains expression in Cast while the variable mapping relationship is created. The unit test validates this scenario.
image

**Comment by fgwang7w** [comment]
@rschlussel when you have a chance, could you please give another round review this fix? Prior review comments have been resolved.  Many thanks

**Comment by rohanpednekar** [comment]
@rschlussel Could you please help to review this PR further?

**Comment by fgwang7w** [comment]
@rschlussel could you please help to have another look? This feature will lift some decorrelation restrictions. thanks

**Comment by rohanpednekar** [comment]
@rschlussel Do you think you can review this PR further? This issue resurfaced again at https://github.com/prestodb/presto/discussions/16010 with Trino distraction.

**Comment by rschlussel** [comment]
sorry for delay. will review

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 3 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:292]: That makes sense. thanks for the explanation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 5]
```diff
@@ -277,29 +278,53 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
...
                 }
 
                 ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                if (!(comparison.getLeft() instanceof SymbolReference
-                        && comparison.getRight() instanceof SymbolReference
-                        && comparison.getOperator().equals(EQUAL))) {
+                // handle coercions and non-direct column references
+                Expression leftExpression;
+                Expression rightExpression;
+                if (comparison.getLeft() instanceof Cast) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:292]: I think this whole code section would be better as an ExpressionVisitor so that it would handle nested expressions and such.

Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java]: yes makes sense, but instead of `ExpressionVisitor`, `AstVisitor` is used to traverse for expression like `Cast` or `SymbolReference`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 6]
```diff
@@ -310,6 +314,22 @@ private boolean isCorrelated(Expression expression)
         {
             return correlation.stream().anyMatch(VariablesExtractor.extractUnique(expression, variableAllocator.getTypes())::contains);
         }
+
+        private class ExpressionEvaluationVisitor
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:318]: I would extend DefaultTraversalVisitor instead and only override visitSymbolReference.  that way you'll automatically have support for nested casts and other kinds of expressions as well.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 7]
```diff
@@ -277,29 +279,31 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
...
-                VariableReferenceExpression right = variableAllocator.toVariableReference(comparison.getRight());
-
-                if (correlation.contains(left) && !correlation.contains(right)) {
-                    mapping.put(left, right);
-                }
-
-                if (correlation.contains(right) && !correlation.contains(left)) {
-                    mapping.put(right, left);
+                // handle coercions and non-direct column references
+                Expression leftExpression = new ExpressionEvaluationVisitor().process(comparison.getLeft());
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:291]: only create the ExpressionEvaluationVisitor once and reuse.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 6 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:318]: sorry that won't quite work

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:318]: maybe do VariablesExtractor.extractAll() or extractUnique() for the right and left sides and then check that each side has exactly one variableReferenceExpression.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 7 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:291]: ok

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 6 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:320]: I replace the current logic of looping through each correlatedConjuncts for correlated symbol mapping by using `VariablesExtractor.extractUnique` to determine any exists in the correlation map. This is a better approach, thanks much for the suggestion!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 8]
```diff
@@ -277,35 +276,40 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
 
         private Multimap<VariableReferenceExpression, VariableReferenceExpression> extractCorrelatedSymbolsMapping(List<Expression> correlatedConjuncts)
         {
-            // TODO: handle coercions and non-direct column references
             ImmutableMultimap.Builder<VariableReferenceExpression, VariableReferenceExpression> mapping = ImmutableMultimap.builder();
             for (Expression conjunct : correlatedConjuncts) {
+                // TODO: Handle non-direct column references for ArithmeticBinaryExpression, ArithmeticUnaryExpression, NotExpression, etc
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:281]: with the new approach for variable extraction do we now support these kinds of expressions?  would be good to add some tests.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java [hunk 9]
```diff
@@ -277,35 +276,40 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
...
-                if (correlation.contains(right) && !correlation.contains(left)) {
-                    mapping.put(right, left);
+                // handle coercions and non-direct column references
+                if (comparison.getOperator().equals(EQUAL)) {
+                    List<VariableReferenceExpression> left = extractUniqueExpression(comparison.getLeft());
+                    List<VariableReferenceExpression> right = extractUniqueExpression(comparison.getRight());
+
+                    for (VariableReferenceExpression leftVariableExpression : left) {
+                        for (VariableReferenceExpression rightVariableExpression : right) {
+                            if (correlation.contains(leftVariableExpression) && !correlation.contains(rightVariableExpression)) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:294]: what does it mean now if there are multiple unique variables on each side of the equals. won't this map the left expression to all fo the right expressions (and similarly for the right code below). that seems incorrect. e.g. what happens if you have `t1.a = t2.a +t2.b`,or ` t1.a + t1.b = t2.a + t2.b`? or even `t1.a + t2.a = t1.b + t2.b`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 8 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:281]: More UT tests added and yes current implementation covers decorrelation on all non-directive column references

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:281]: did you mean to update the PR? don't see any changes.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 9 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java]: when there are multiple unique column variables on each side of the comparison expression, they need to be validated if each unique variable from each side is a correlated variable and can be mapped to the right side's unique variable. 

E.g. `t1.a = t2.a  + t2.b` is tranformed into 
`cast(t1.a) = t2.a+t2.b` which would have a  a correlation mapping variables of `t2.a` and `t2.b` and a `cast(t1.a)`. 
So now left variable `t1.a` is in a mappting to each of right side variables `t2.a` and `t2.b`.

For `t1.a + t2.a = t1.b + t2.b`, the correlated predicate is tranformed into 
`cast(t1.a) + t2.a = cast(t1.b) + t2.b` for which the correlation variables are `t2.a`, `t2.b` with `cast(t1.a)` and `cast(t1.b)`.
so the mapping becomes `t2.b`-> `t1.a1` and `t2.a`->`t1.b`
image

I also added UT test coverage on some cases mentioned here to resolve the confusion.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 8 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java]: yes PR is updated now, I was cleaning up some comments and UTs

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java [hunk 10]
```diff
@@ -118,15 +118,35 @@ public void testCorrelatedSubqueriesWithLimit()
...
+        // Test decorrelated subexpression with ArithmeticBinaryExpression
+        assertions.assertQuery(
+                "select (select count(*) from (values 1, 1, 3) t(a) where t.a = t2.a + t2.b group by a limit 1) from (values (1.0, 1.0),(1.0,2.0)) t2(a,b)",
+                "VALUES null, BIGINT '1'");
+        // Test decorrelated subexpression with ArithmeticUnaryExpression
+        assertions.assertQuery(
+                "select (select count(*) from (values 1, 2, 3,4,5) t(a) where t.a = t2.a + (-t2.b)+1 group by a limit 1) from (values (1.0, 1.0),(1.0, 2.0),(1.0,2.0)) t2(a,b)",
+                "VALUES null, null, BIGINT '1'");
+        // Test decorrelated subexpression with NotExpression
+        assertions.assertQuery(
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:141]: I think this needs an order by to be deterministic.  if it's `ORDER BY a DESC`, then the answer is `1` (the row where a=3), but if it's `ORDER BY a ASC`, then the answer is `2`  (the two rows where a=1).  It would better though to leave out the group by + limit.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java [hunk 11]
```diff
@@ -86,16 +86,16 @@ public void testCorrelatedExistsSubqueriesWithOrPredicateAndNull()
...
     @Test
-    public void testUnsupportedSubqueriesWithCoercions()
+    public void testSubqueriesWithCoercions()
     {
         // coercion from subquery symbol type to correlation type
-        assertions.assertFails(
-                "select (select count(*) from (values 1) t(a) where t.a=t2.b limit 1) from (values 1.0) t2(b)",
-                UNSUPPORTED_CORRELATED_SUBQUERY_ERROR_MSG);
+        assertions.assertQuery(
+                "select (select count(*) from (values 1) t(a) where t.a=t2.b group by a limit 1) from (values 1.0) t2(b)",
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:93]: why does this need a group by? or limit for that matter

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java [hunk 12]
```diff
@@ -118,15 +118,35 @@ public void testCorrelatedSubqueriesWithLimit()
...
                 "VALUES true",
                 false);
+        // Test decorrelated subexpression with ArithmeticBinaryExpression
+        assertions.assertQuery(
+                "select (select count(*) from (values 1, 1, 3) t(a) where t.a = t2.a + t2.b group by a limit 1) from (values (1.0, 1.0),(1.0,2.0)) t2(a,b)",
+                "VALUES null, BIGINT '1'");
+        // Test decorrelated subexpression with ArithmeticUnaryExpression
+        assertions.assertQuery(
+                "select (select count(*) from (values 1, 2, 3,4,5) t(a) where t.a = t2.a + (-t2.b)+1 group by a limit 1) from (values (1.0, 1.0),(1.0, 2.0),(1.0,2.0)) t2(a,b)",
+                "VALUES null, null, BIGINT '1'");
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:139]: I will add a flag to ensure ordering is checked as well. format is fixed too.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 11 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:93]: this was from original sql statement, but on second, we don't need it since the result is a count aggregation.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 10 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:141]: Are you suggesting to do `ORDER BY` inside the subquery? The ordering is unnecessary because the query returns a final count. sure I will leave out the groupby + limit clause

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java (hunk 4 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java:280]: Marking that with the latest implementation, we won't have any other non-directive column reference decorrelation issues

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 10 above)
Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:143]: The ordering is necessary because the query returns a count grouped by a, and there are 2 values of a.  Without the ordering, it's not deterministic which count will be returned.    By the way, I think I meant just leave out the limit, though leaving out both group by and limit is okay too.  Also, the outermost select is not necessary in this query.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java]: no need for the flag, or if you do add the flag the query needs an order by (otherwise order is not guaranteed).  I just meant that for clarity of understanding the results, it would be better to list the output rows in the same order as the input rows.

[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java]: Undestood, I will remove the flag but keep the output in the right order

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 10 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:143]: I see, that's a good point, thank you, I will fix the ut now with `order by` clause to make it clear

Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:141]: The order by isn't needed anymore since you removed the limit.  It was only needed so it could be deterministic which one value would be returned.

Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:141]: apology for the confusion. I took out the `order by` first and then added it back but left out limit. Anyway, I'm going to add `order by asc` and `order by desc` both cases

**Comment by fgwang7w** [comment]
Hi @rschlussel all comments are resolved for this PR, could you please kindly take a look again? many thanks

**Review by rschlussel — APPROVED** [review]
